packages/config/src/Devices.ts

```diff
@@ -41,66 +41,107 @@ export const devicesDir = path.join(configDir, "devices");
 export const indexPath = path.join(devicesDir, "index.json");
 export type DeviceConfigIndex = DeviceConfigIndexEntry[];
 
-/** @internal */
-export async function loadDeviceIndexInternal(): Promise<DeviceConfigIndex> {
-	if (!(await pathExists(indexPath))) {
-		return generateDeviceIndex();
-	}
-	try {
-		const fileContents = await readFile(indexPath, "utf8");
-		return JSON5.parse(fileContents);
-	} catch (e: unknown) {
-		if (e instanceof ZWaveError) {
-			throw e;
-		} else {
-			throwInvalidConfig("devices");
+async function hasChangedDeviceFiles(
+	dir: string,
+	lastChange: Date,
+): Promise<boolean> {
+	// Check if there are any files BUT index.json that were changed
+	// or directories that were modified
+	const filesAndDirs = await fs.readdir(dir);
+	for (const f of filesAndDirs) {
+		const fullPath = path.join(dir, f);
+
+		const stat = await fs.stat(fullPath);
+		if (
+			(dir !== devicesDir || f !== "index.json") &&
+			(stat.isFile() || stat.isDirectory()) &&
+			stat.mtime > lastChange
+		) {
+			return true;
+		} else if (stat.isDirectory()) {
+			// we need to go deeper!
+			if (await hasChangedDeviceFiles(fullPath, lastChange)) return true;
 		}
 	}
+	return false;
 }
 
-/** Generates an index for all device config files */
-async function generateDeviceIndex(): Promise<DeviceConfigIndex> {
-	const configFiles = await enumFilesRecursive(
-		devicesDir,
-		(file) => file.endsWith(".json") && !file.endsWith("index.json"),
-	);
+/**
+ * @internal
+ * Loads the index file to quickly access the device configs.
+ * Transparently handles updating the index if necessary
+ */
+export async function loadDeviceIndexInternal(
+	forceWrite?: boolean,
+): Promise<DeviceConfigIndex> {
+	// The index file needs to be regenerated if it does not exist
+	let needsUpdate = !(await pathExists(indexPath));
+	let index: DeviceConfigIndex | undefined;
+	let mtimeIndex: Date | undefined;
+	// ...or if cannot be parsed
+	if (!needsUpdate) {
+		try {
+			const fileContents = await readFile(indexPath, "utf8");
+			index = JSON5.parse(fileContents);
+			mtimeIndex = (await fs.stat(indexPath)).mtime;
+		} catch {
+			// console.error("Error while parsing index file - regenerating...");
```

Comment:
Remove console.error or replace them with logger errors

Reply:
that's commented out, but I can remove it ;)

Reply:
I would use the logger here too (if present), fallback to console if not